core/auth/src/main/java/software/amazon/awssdk/auth/credentials/SessionCredentialsHolder.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.auth.credentials;
+
+import java.time.Instant;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.annotations.ThreadSafe;
+
+/**
+ * Holder class used to atomically store a session with its expiration time.
+ */
+@SdkProtectedApi
+@ThreadSafe
+public final class SessionCredentialsHolder {
```

Comment:
I know it hurts to duplicate this, but I'd vote for copying it to the two places it's needed instead of sharing it, and making it an API we have to maintain backwards-compatibility on.

Maybe an alternative would be to make AwsSessionCredentials have an optional expiration time? Not sure if that's better or worse...

Reply:
Can I know the reason why we don't want to share this between different credentials providers?

Reply:
Our biggest source of technical debt in the SDK are APIs that we can't change: protected and public APIs. We try not to add new protected or public APIs unless we have a strong reason to do so, because it means we can pretty much never change our mind in the future. We can always refactor internal APIs if we want to in the future.

This seems like a protected API that we might regret adding. For example, what happens if we want to add the expiration time to AwsSessionCredentials in the future? Then, SessionCredentialsHolder is no longer needed, but we can't delete it because it's protected.

Reply:
Got you, will do the change.